### ISSUE TITLE
Script API: support pixel format selection for DynamicSprites

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#2027]
**Problem**

Currently AGS converts any sprite or loaded image at runtime to the native game color depth. This may be fine in the majority of cases, but there are still situations where you may want that to not happen.

First of all, user may like to import a 8-bit image to use as a mask in their game (in fact, they may even want 4, 2 or 1-bit mask, something that AGS never supported directly).

Another, perhaps less important, case, is if somebody will like to have a strictly opaque surface and thus have a 24-bit RGB sprite, to ensure that it's always stays opaque regardless of what you draw on it.
In the recent past I've removed the "has alpha channel" flag (#1808, #1813), because it was creating unnecessary difficulties when handling sprites, but in this case the pixel format of the bitmap itself will affect the draw/blend result, so no extra logic in the engine will be necessary (everything is handled by the bitmap vtable implementations).
Allegro bitmap impl already supports 24-bit, and that may be even replaced by a custom 32-bit RGB implementation that keeps alpha channel opaque (0xFF) at all times.
This part is not a priority here though, and may be omitted.

**Proposal**

We need to make sure that the engine does not do implicit conversions unless user lets it. The solution I'm proposing is to add an 
*optional* argument to the DynamicSprite.Create* and LoadFromFile functions, that tells the wanted pixel format. In Create* functions this argument will take place of a recently removed "has alpha" argument, and we may actually pick constant values that match the "has alpha" values for easier transition.

I'm thinking about something like

```
enum PixelFormat {
    // use default game's format for new sprites, or keep source image's format if possible
    ePixelFmt_Default = -1,
    // RGB in game's format (if supported, otherwise - default game's format, e.g. indexed 8-bit)
    ePixelFmt_RGB = 0,
    // ARGB in game's format (if supported, otherwise - see RGB)
    ePixelFmt_ARGB = 1,
    // 8-bit indexed format
    ePixelFmt_Indexed8bit = 2,
    // this may be extended further
};
```

The proposed look of the DynamicSprite functions then will be:
```
DynamicSprite* Create(int width, int height, PixelFormat fmt = ePixelFmt_ARGB);
DynamicSprite* CreateFromExistingSprite(int slot, PixelFormat fmt = ePixelFmt_Default);
DynamicSprite* CreateFromFile(const string filename, PixelFormat fmt = ePixelFmt_Default);
```

Add a new readonly property `readonly PixelFormat DynamicSprite.Format`, for informational purposes.

**Comment by ivan-mogilko** [comment]
There's a related problem with the sprites imported into the game project. The idea is to let import, for example, a 8-bit sprite for the purpose of using as a mask, and keep its format.

What would this require?:
- a sprite import option; may be a switch between 2 variants: convert to game default format, or keep original.
- on the engine side, need to make sure it is possible to distinguish sprites that has to be brought to game default fmt, and ones that do not.

Perhaps I will write another ticket for this.

**Comment by ericoporto** [comment]
About importing 8-bit sprites, I am not sure, but I believe in 8-bit games they don't store the palette in the sprite file.

**Comment by ivan-mogilko** [comment]
> About importing 8-bit sprites, I am not sure, but I believe in 8-bit games they don't store the palette in the sprite file.

No they don't, although spritefile format has a place for palette, but it's left empty, and engine merges two palettes from game and current room data instead.

**Comment by ericoporto** [comment]
Hey, I am looking at the enum, and I am think what if we left out both the values for 0 and 1? Since the position for the parameter in the function previously had a boolean there, in this way any script that doesn't break when compiling wouldn't not give a weird behavior - if a person forgot to update the script.

**Comment by ivan-mogilko** [comment]
> or perhaps ePixelFmt_RGB is false opaque and ePixelFmt_ARGB is transparency and that is why they are 0 and 1?

Yes, that was the idea.

**Comment by ivan-mogilko** [comment]
I've been looking into this recently, and had a slight change in opinion about the necessary options.

First of all, I now think that having a special "PixelFormat" option is not necessary, as that might be confusing to users. Having special enums for "default RGB" and "default ARGB" looks redundant.
Instead, it may make more sense to have a more direct "Color Depth" parameter, simply as integer (e.g. 8, 24, 32), where value "0" would mean "use game's default". If really wanted, we could still make enum, but this enum's constant's would match these values; for example:  `ePixelFmt_Indexed8bit = 8`.

Then, I realized that having same "color depth" option for CreateFromFile is going to be superfluous, because when loading a sprite there really are only 2 meaningful options:
* keep source's color depth (if supported)
* convert to game default color depth
On another hand, there may be additional options; for example: when loading a 8-bit image we may want to keep original color indexes (mask case) or convert to game's palette (regular sprite case).

This brought me to idea that CreateFromFile should have the same options as "sprite import" in the Editor; at least the basic ones:
* convert to game's color depth or keep original
* remap palette or no
* choose transparent pixel

In other words, have enum like ImageLoadFlags with a set of bit flags in it.

At last, CreateFromExistingSprite again does not need a "pixel format" option, but either have same option as CreateFromFile, or no extra option at all. The most common use case for making a sprite's copy is to modify it. In which case we likely want to keep source format. If changing color depth is necessary for some reason, then the solution would be to create a new sprite of chosen color depth, and paint previous sprite on it. The only meaningful use case I can imagine here is when you have a 8-bit sprite used as a mask (for example) which you like to paint on a 32-bit surface in order to display somewhere.

**Comment by ericoporto** [comment]
I think I would avoid exposing for user writing the color format as integer and use an enum instead because in reality the color format is both how channels are arranged and the bitness - in theory you could have ARGB with 6-bit per channel in 24-bit.

There is an issue that a surface has a DrawImage command and I am not sure what happens when one draws something from whatever format to a surface in a different format. Also I am not sure what happens with COLOR_TRANSPARENT here.

**Comment by ivan-mogilko** [comment]
> I think I would avoid exposing for user writing the color format as integer and use an enum instead because in reality the color format is both how channels are arranged and the bitness - in theory you could have ARGB with 6-bit per channel in 24-bit.

My point is that user does not need to know these specifics. The actual pixel format is chosen by the engine as most suitable for the given color depth, same way as it choses pixel format from the game's color depth.

> There is an issue that a surface has a DrawImage command and I am not sure what happens when one draws something from whatever format to a surface in a different format

This is already handled by having blit with conversion.

**Comment by ivan-mogilko** [comment]
Resolved by #2455
